File: banco.php
<?php
// Clase Usuario que representa a un cliente del banco
class Usuario
{
    public $nombre;
    public $pin;
    public $saldo;

    public function __construct($nombre, $pin, $saldo)
    {
        $this->nombre = $nombre;
        $this->pin = $pin;
        $this->saldo = $saldo;
    }
}

// Clase Banco que maneja los usuarios y sus operaciones
class Banco
{
    private $usuarios = [];

    public function __construct($usuarios = [])
    {
        $this->usuarios = $usuarios;
    }

    // Método para autenticar al usuario
    public function autenticar($nombre, $pin)
    {
        foreach ($this->usuarios as $usuario) {
            if ($usuario->nombre === $nombre) { // Verificamos si el usuario existe
                if ($usuario->pin == $pin) { // Comprobamos si el PIN es correcto
                    echo "¡Estás logueado!\n";
                    return true;
                } else {
                    echo "Pin o nombre incorrecto.\n";
                    return false;
                }
            }
        }
        echo "El usuario no existe.\n";
        return false;
    }

    // Método para retirar dinero de la cuenta de un usuario
    public function sacarDinero($nombre, $cantidad)
    {
        foreach ($this->usuarios as $usuario) {
            if ($usuario->nombre === $nombre) {
                if ($usuario->saldo < $cantidad) { // Verificamos si hay saldo suficiente
                    echo "Saldo insuficiente.\n";
                } else {
                    $usuario->saldo -= $cantidad; // Restamos el monto al saldo
                    echo "El saldo disponible es: " . $usuario->saldo . "\n";
                }
                return;
            }
        }
    }
}

// Crear instancias de usuarios con nombre, PIN y saldo inicial
$ana = new Usuario("Ana", 9872, 450);
$pablo = new Usuario("Pablo", 5555, 200);
$rodrigo = new Usuario("Rodrigo", 2222, 900);

// Crear una instancia del Banco con los usuarios registrados
$banco = new Banco([$ana, $pablo, $rodrigo]);

// Función para solicitar entrada del usuario
function solicitarEntrada($mensaje)
{
    echo $mensaje;
    return trim(fgets(STDIN));
}

// Función para iniciar sesión en el sistema
function iniciarSesion($banco)
{
    $nombre = solicitarEntrada("Introduzca su nombre: ");
    $pin = solicitarEntrada("Introduzca su pin: ");

    if (!is_numeric($pin)) {
        echo "El PIN debe ser un número.\n";
        iniciarSesion($banco);
        return;
    }

    if ($banco->autenticar($nombre, (int)$pin)) {
        menuUsuario($banco, $nombre);
    } else {
        iniciarSesion($banco);
    }
}

// Función que muestra el menú de opciones para el usuario autenticado
function menuUsuario($banco, $nombre)
{
    echo "Por favor, elija una opción:\n1. Sacar dinero\n2. Terminar sesión\n";
    $opcion = solicitarEntrada("");

    if ($opcion === "1") { // Opción para retirar dinero
        $cantidad = solicitarEntrada("Introduce cantidad a sacar: ");

        if (!is_numeric($cantidad) || (int)$cantidad <= 0) {
            echo "Por favor, ingrese un monto válido.\n";
            menuUsuario($banco, $nombre);
            return;
        }

        $banco->sacarDinero($nombre, (int)$cantidad);
        menuUsuario($banco, $nombre); // Volvemos a mostrar el menú

    } elseif ($opcion === "2") { // Opción para cerrar sesión
        echo "Saliendo del sistema.\n";
        exit;
    } else { // Opción inválida
        echo "Opción incorrecta.\n";
        menuUsuario($banco, $nombre);
    }
}

// Mensaje de bienvenida e inicio de sesión
echo "Bienvenido al Banco\n";
iniciarSesion($banco);

File: C mas/banco.cpp
#include <iostream>
#include <vector>
using namespace std;

// Clase que representa un usuario del banco
class Usuario
{
public:
    string nombre; // Nombre del usuario
    int pin;       // PIN de acceso
    double saldo;  // Saldo de la cuenta

    // Constructor para inicializar los atributos del usuario
    Usuario(string n, int p, double s)
    {
        nombre = n;
        pin = p;
        saldo = s;
    }
};

// Clase que representa el banco y sus operaciones
class Banco
{
private:
    vector<Usuario> usuarios; // Lista de usuarios

public:
    // Método para agregar usuarios al banco
    void agregarUsuario(Usuario usuario)
    {
        usuarios.push_back(usuario);
    }

    // Método para autenticar un usuario por nombre y PIN
    Usuario *autenticar(string nombre, int pin)
    {
        for (auto &usuario : usuarios)
        {
            if (usuario.nombre == nombre)
            {
                if (usuario.pin == pin)
                {
                    cout << "✅ Autenticación exitosa." << endl;
                    return &usuario;
                }
                else
                {
                    cout << "❌ PIN incorrecto." << endl;
                    return nullptr;
                }
            }
        }
        cout << "❌ El usuario no existe." << endl;
        return nullptr;
    }

    // Método para retirar dinero del usuario autenticado
    void retirarDinero(Usuario *usuario, double cantidad)
    {
        if (usuario->saldo >= cantidad)
        {
            usuario->saldo -= cantidad;
            cout << "✅ Retiro exitoso. Nuevo saldo: $" << usuario->saldo << endl;
        }
        else
        {
            cout << "❌ Saldo insuficiente. Saldo disponible: $" << usuario->saldo << endl;
        }
    }

    // Método para consultar el saldo del usuario
    void consultarSaldo(Usuario *usuario)
    {
        cout << "💰 Saldo actual: $" << usuario->saldo << endl;
    }
};

int main()
{
    Banco banco;

    // Se agregan usuarios al banco
    banco.agregarUsuario(Usuario("ana", 9872, 450));
    banco.agregarUsuario(Usuario("pablo", 5555, 200));
    banco.agregarUsuario(Usuario("rodrigo", 2222, 900));

    while (true)
    {
        cout << "\n💰 Bienvenido al banco. Por favor, identifíquese." << endl;
        Usuario *usuario = nullptr;
        string nombre;
        int pin, intentos = 3;

        while (intentos > 0 && usuario == nullptr)
        {
            cout << "Introduzca su nombre: ";
            cin >> nombre;
            cout << "Introduzca su PIN: ";
            cin >> pin;

            usuario = banco.autenticar(nombre, pin);

            if (usuario == nullptr)
            {
                intentos--;
                if (intentos == 0)
                {
                    cout << "⛔ Demasiados intentos fallidos. Volviendo al inicio..." << endl;
                }
            }
        }

        if (usuario != nullptr)
        {
            while (true)
            {
                cout << "\n📌 Opciones: 1. Retirar dinero  2. Consultar saldo  3. Salir\n";
                string opcion;
                cin >> opcion;

                if (opcion == "1")
                {
                    double cantidad;
                    cout << "💵 Cantidad a retirar: ";
                    cin >> cantidad;
                    banco.retirarDinero(usuario, cantidad);
                }
                else if (opcion == "2")
                {
                    banco.consultarSaldo(usuario);
                }
                else if (opcion == "3")
                {
                    cout << "👋 Sesión cerrada.\n";
                    break;
                }
            }
        }
    }
    return 0;
}
